### PR TITLE
Try to avoid etcd.Get as part of Delete operation

### PR DESCRIPTION
**Pull request by wojtek-t**
This was done for GuaranteedUpdate before in #35415 but later there were multiple (direct or indirect) bug fixes to that logic:
https://github.com/kubernetes/kubernetes/pull/40664 : Allow values to be wrapped prior to serialization in etcd
https://github.com/kubernetes/kubernetes/pull/47703 : Do not persist SelfLink into etcd storage
https://github.com/kubernetes/kubernetes/pull/48394 : GuaranteedUpdate must write if stored data is not canonical
https://github.com/kubernetes/kubernetes/pull/43152 : etcd3 store: retry with live object on conflict if there was a suggestion
https://github.com/kubernetes/kubernetes/pull/54780 : partial fix crd patch failing
https://github.com/kubernetes/kubernetes/pull/58375 : Recheck if transformed data is stale when doing live lookup during update
https://github.com/kubernetes/kubernetes/pull/77619 : In GuaranteedUpdate, retry on any error if we are working with cached data
https://github.com/kubernetes/kubernetes/pull/78713 : Set expected in-memory version when decoding unstructured objects from etcd
https://github.com/kubernetes/kubernetes/pull/82303 : In GuaranteedUpdate, retry on a precondition check failure if we are working with cached data

The PR upfront tries to address it by extensive testing based on all the issues listed above

Depending on the run and scale, we've seen between 10% and 60% reduction of latency on 99th percentiled for Delete API calls.

```release-note
NONE
```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/storage/interfaces.go [hunk 1]
```diff
@@ -166,7 +166,11 @@ type Interface interface {
 
 	// Delete removes the specified key and returns the value that existed at that spot.
 	// If key didn't exist, it will return NotFound storage error.
-	Delete(ctx context.Context, key string, out runtime.Object, preconditions *Preconditions, validateDeletion ValidateObjectFunc) error
+	// If 'suggestion' can contain zero or one element - in such case this can be used as
+	// a suggestion about the current version of the object to avoid read operation from
+	// storage to get it.
+	Delete(ctx context.Context, key string, out runtime.Object, preconditions *Preconditions,
+		validateDeletion ValidateObjectFunc, suggestion ...runtime.Object) error
```
Comment (enj) [inline staging/src/k8s.io/apiserver/pkg/storage/interfaces.go:173]: Please use `suggestion runtime.Object`.  The existing anti-pattern of using `suggestion ...runtime.Object` makes is unnecessarily hard to find callers that give a suggestion vs. not.  Go does not have optional parameters to functions; please do not try to invent such an approach.

Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/storage/interfaces.go:173]: We can consider cleaning this up, but I strongly believe consistency is much more important. Given we already have it for Update, as a user I don't want to have two functions that are supposed to have the same semantic and are using a different interface...

**Comment by liggitt** [comment]
Do you have evidence that the performance gains are significant? this makes the code quite a bit more complex, and it's not as obvious that deleting from cache is as much of a performance gain as updating from cache. We had many regressions because of update-from-cache, and I'm not confident our test coverage here is as good as it needs to be.

**Comment by liggitt** [comment]
> I'm not confident our test coverage here is as good as it needs to be.

For example, one thing I started doing was [running tests with watch cache both on and off](https://grep.app/search?q=With%28out%29%3FWatchCache&regexp=true&filter[repo][0]=kubernetes/kubernetes) to verify identical behavior.

**Comment by wojtek-t** [comment]
> Do you have evidence that the performance gains are significant? this makes the code quite a bit more complex, and it's not as obvious that deleting from cache is as much of a performance gain as updating from cache.

Sure - that's a good question. All debugging shows it will help a lot, but we need to run scale tests to provide numbers. Holding before we have them.
/hold

> We had many regressions because of update-from-cache, and I'm not confident our test coverage here is as good as it needs to be.

We used to have a bunch in the early days, but I'm not sure I'm aware of any in the last 2 years or so. I also think that it's not adding a completely new code-path per-se - we are duplicating something that already exists for 3+ years, so I think that is quite different thing.

> For example, one thing I started doing was running tests with watch cache both on and off to verify identical behavior, and found multiple bugs in the process.

Hmm - can you point me to them? I would be very interested in seeing what can go wrong.

**Comment by liggitt** [comment]
I did a quick sweep of bugs just in the last few releases that were triggered by using the cached object in Update operations:
* this month: bug in decoder, only symptom was unconverted object in watch cache on LIST request:
  * https://github.com/kubernetes/kubernetes/issues/88473
  * https://github.com/kubernetes/kubernetes/pull/88995
* 1.16: precondition comparison failure with old object
  * https://github.com/kubernetes/kubernetes/issues/82130#issuecomment-527665796
  * https://github.com/kubernetes/kubernetes/pull/82303
* 1.15: another decoding bug, only symptom was unconverted object in watch cache from WATCH events:
  * https://github.com/kubernetes/kubernetes/issues/72493#issuecomment-498861554
  * https://github.com/kubernetes/kubernetes/pull/78713

Example tests added for in-memory updates with/without watch cache enabled (we have automated tests like this for a few operations, but not most):
* https://github.com/kubernetes/kubernetes/pull/79659

Example test artifically delaying the watch cache to trigger bugs from stale objects (we don't have automated tests like this):
* https://github.com/kubernetes/kubernetes/issues/82130#issuecomment-527683188

**Comment by wojtek-t** [comment]
/remove-lifecycle stale

I'm going to get back to it.

/lifecycle frozen

File: staging/src/k8s.io/apiserver/pkg/storage/interfaces.go (hunk 1 above)
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/storage/interfaces.go:173]: done

Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/storage/interfaces.go]: done

**Comment by wojtek-t** [comment]
Neither of the errors couldn't be caused by problems with delete.
/retest

**Comment by wojtek-t** [comment]
@liggitt - I have analyzed deeply everything that happened with GuaranteedUpdate, added some tests to upfronet ensure that issues that it caused are tested explicitly (for those that it made sense) and I think it's ready for a pass of review; PTAL when you will get out of your 1.20 work

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go [hunk 2]
```diff
@@ -227,8 +269,13 @@ func (s *store) conditionalDelete(ctx context.Context, key string, out runtime.O
 			return err
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:269]: add a unit test to make sure the right thing happens if the suggestion is stale and the key no longer exists (was already deleted). I think we would get a NotFound here and return, but verify the right thing happens

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go [hunk 3]
```diff
@@ -431,8 +431,19 @@ func (c *Cacher) Create(ctx context.Context, key string, obj, out runtime.Object
...
-	return c.storage.Delete(ctx, key, out, preconditions, validateDeletion)
+func (c *Cacher) Delete(
+	ctx context.Context, key string, out runtime.Object, preconditions *storage.Preconditions,
+	validateDeletion storage.ValidateObjectFunc, suggestion runtime.Object) error {
+	// Ignore the suggestion and try to pass down the current version of the object
+	// read from cache.
+	if elem, exists, err := c.watchCache.GetByKey(key); err != nil {
+		klog.Errorf("GetByKey returned error: %v", err)
+	} else if exists {
+		currObj := elem.(*storeElement).Object.DeepCopyObject()
```
[the remark this answers is not in the document]
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:444]: done

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go [hunk 4]
```diff
@@ -185,35 +185,77 @@ func (s *store) Create(ctx context.Context, key string, obj, out runtime.Object,
...
+	return s.conditionalDelete(ctx, key, out, v, preconditions, validateDeletion, suggestion)
 }
 
-func (s *store) conditionalDelete(ctx context.Context, key string, out runtime.Object, v reflect.Value, preconditions *storage.Preconditions, validateDeletion storage.ValidateObjectFunc) error {
-	startTime := time.Now()
-	getResp, err := s.client.KV.Get(ctx, key)
-	metrics.RecordEtcdRequestLatency("get", getTypeName(out), startTime)
+func (s *store) conditionalDelete(
+	ctx context.Context, key string, out runtime.Object, v reflect.Value, preconditions *storage.Preconditions,
+	validateDeletion storage.ValidateObjectFunc, suggestion runtime.Object) error {
```
[the remark this answers is not in the document]
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go]: done

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go [hunk 5]
```diff
@@ -185,35 +185,77 @@ func (s *store) Create(ctx context.Context, key string, obj, out runtime.Object,
...
+		metrics.RecordEtcdRequestLatency("get", getTypeName(out), startTime)
+		if err != nil {
+			return nil, err
+		}
+		return s.getState(getResp, key, v, false)
+	}
+
+	var origState *objState
+	var err error
+	var mustCheckData bool
```
[the remark this answers is not in the document]
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go]: done

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go (hunk 2 above)
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:269]: Added "TestDeleteWithSuggestionOfDeletedObject" test.

That said - it doesn't exercise this path. What happens is that the transation fails, we get into
```
if !txnResp.Succeeded {
```
branch and this is handled in getState (getResp contains empty KV field and that is handled as returning "not-found" error).

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go [hunk 6]
```diff
@@ -185,35 +185,77 @@ func (s *store) Create(ctx context.Context, key string, obj, out runtime.Object,
 }
 
 // Delete implements storage.Interface.Delete.
-func (s *store) Delete(ctx context.Context, key string, out runtime.Object, preconditions *storage.Preconditions, validateDeletion storage.ValidateObjectFunc) error {
+func (s *store) Delete(
```
[the remark this answers is not in the document]
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:188]: Done.
I added one more thing to `TestValidateDeletionWithSuggestion` test.
The remaining uncovered path are error handling from etcd layer (but this is simple returning on error - those paths:
```
if err != nil {
  return err
}
```

Other than everything is covered.

**Review by wojtek-t — COMMENTED** [review]
@liggitt - comments addressed PTAL

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go (hunk 2 above)
Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:269]: `txnResp.Responses[0]` exists for a NotFound response?

**Comment by liggitt** [comment]
ok, I'm satisfied this is functionally correct. I still didn't see the performance numbers demonstrating the benefit was worth the additional complexity:

>> Do you have evidence that the performance gains are significant? this makes the code quite a bit more complex, and it's not as obvious that deleting from cache is as much of a performance gain as updating from cache.
>
> Sure - that's a good question. All debugging shows it will help a lot, but we need to run scale tests to provide numbers. Holding before we have them.

Did I miss where those were provided?

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go (hunk 2 above)
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:269]: Yes - it exists and being GetResponse contains empty set of KV pairs.

**Comment by wojtek-t** [comment]
> Did I miss where those were provided?

Sorry - I forgot to add that in the meantime. I added the following to the PR description:
`Depending on the run and scale, we've seen between 10% and 60% reduction of latency on 99th percentiled for Delete API calls.`

@liggitt - PTAL

**Pull request merged**